### PR TITLE
feat(nnx): support native Flax NNX PEFT/LoRA training loop

### PR DESCRIPTION
**Pull request by RexBearIU**
# Description

This PR adds native parameter-efficient fine-tuning (PEFT), LoRA, and QLoRA training support and complete checkpoint lifecycle resume/restoration within Flax NNX training in MaxText (`train.py`, `train_sft.py`, `maxtext_utils.py`, `checkpointing.py`, `train_state_nnx.py`, `lora_utils.py`, `sharding.py`).

### Key Changes:

1. **Dynamic Parameter Filtering (`wrt`)**:
   - Standardized `nnx.Optimizer(wrt=...)` across MaxText (`train.py`, `train_sft.py`, `train_compile.py`, `generate_param_only_checkpoint.py`, `standalone_checkpointer.py`, `train_utils.py`) to target `nnx.LoRAParam` when LoRA is enabled, ensuring gradient updates affect adapter weights.

2. **Mesh Sharding & Activation Protection**:
   - Scoped `adjust_pspec_for_indivisible_shapes` in `src/maxtext/utils/sharding.py` strictly to parameter variables (`isinstance(inputs, nnx.Variable)`), preserving activation partition specs inside `shard_map` across multi-host TPU topologies.

3. **Checkpoint Lifecycle & Resume**:
   - Integrated Flax NNX checkpoint serialization and restoration in `src/maxtext/common/checkpointing.py`, supporting both Linen `'params'` wrappers and root NNX state dictionaries.
   - Updated `TrainStateNNX.split_for_checkpoint` in `train_state_nnx.py` to route `nnx.Param` and `nnx.LoRAParam` into parameter collections for complete PyTree state tracking.

4. **QWiX Quantization & Scanned Layer Execution**:
   - Added safe kernel unwrapping (`getattr(self.kernel, "value", self.kernel)`) in `src/maxtext/layers/linears.py` for QWiX quantized parameter wrappers.
   - Bypassed `nnx.split`/`nnx.merge` for QWiX layers in `src/maxtext/layers/nnx_decoders.py` and filtered `RngState` and `Intermediate` variables during layer scanning.

5. **SFT Mesh Placement & Imports**:
   - Wrapped model initialization and adapter restoration in `train_sft.py` within `with jax.set_mesh(mesh)` to ensure multi-device mesh placement.
   - Lazily imported `train_sft` in test helpers and decorated SFT test methods with `@pytest.mark.post_training`.

# Tests

- **Pre-Commit**: Passed 100% (`pyink`, `pylint`, `codespell`, `yamllint`).
- **Pipeline Parallelism**: All 6 tests in `tests/integration/pipeline_parallelism_test.py` passed (`test_full_train_circular`, `test_full_train_non_circular`, `test_subset_layers`, `test_full_train_fp8`, `test_full_train_nanoo_fp8`).
- **NNX Unit Tests**: `tests/unit/maxtext_utils_nnx_test.py` and `tests/unit/lora_utils_nnx_test.py` passed 100%.
- **E2E Integration Tests**: E2E QLoRA/LoRA tests (`test_qlora_e2e_gemma4_sft_nf4`, `test_qlora_e2e_qwen3_sft_nf4`, `test_qlora_e2e_gptoss_unscanned_nf4`) passed 100%.

# Checklist

- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests and provided workload details above.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: src/maxtext/utils/train_utils.py [hunk 1]
```diff
@@ -241,7 +242,11 @@ def setup_train_loop(config, recorder, devices=None):
       # For NNX, the train state is wrapped in the TrainStateNNX module.
       def create_train_state_fn():
         model = _create_model_partial()
-        optimizer = nnx.Optimizer(model, tx, wrt=nnx.Param)
+        wrt = nnx.Param
+        if config.lora.enable_lora:
```
Comment (SurbhiJainUSC) [inline src/maxtext/utils/train_utils.py]: nit: `getattr(getattr(config, "lora", None), "enable_lora", False)`

Comment (SurbhiJainUSC) [reply src/maxtext/utils/train_utils.py]: Make sure this is uniform across all files to ensure safe configuration check.

Comment (RexBearIU) [reply src/maxtext/utils/train_utils.py]: Done! Made the config check uniform across all relevant modified files in the change list using the safer nested getattr pattern: `getattr(getattr(config, "lora", None), "enable_lora", False)`.

**Review by SurbhiJainUSC — APPROVED** [review]

File: src/maxtext/utils/train_utils.py (hunk 1 above)
Comment (RexBearIU) [reply src/maxtext/utils/train_utils.py]: I have pushed a fix that resolves the `UnexpectedTracerError` during `eval_shape` when restoring checkpoints with LoRA enabled. All integration tests (`peft_integration_test.py` and `setup_train_loop_nnx_test.py`) are now passing.

**Review by YixuanWang-99 — APPROVED** [review]

**Comment by RexBearIU** [comment]
Merged via commit 8fcc9e133da3ef6e6b46e59121bb318fb6f1db36
